calyx/src/analysis/order_analysis.rs

```diff
@@ -0,0 +1,151 @@
+        // checks 1'd1
+        && asmt.src.borrow().is_constant(1, 1)
+    }
+
+    /// Wheters whether amt writes to a component or non-combinational primitive,
+    /// *or* writes to the group's done port.
+    pub fn is_expected_write(asmt: &ir::Assignment) -> bool {
+        match &asmt.dst.borrow().parent {
+            ir::PortParent::Cell(cell) => comp_or_non_comb(&cell.upgrade()),
+            ir::PortParent::Group(_) => asmt.dst.borrow().name == "done",
+        }
+    }
+
+    // For a given asmt, if asmt is a.go = b.done, then we add (b,a) to self.go_done_map.
+    // Also if asmt is group[done] = cell.done, sets self.last to Some(cell).
+    fn update(&mut self, asmt: &ir::Assignment) {
+        let src = asmt.src.borrow();
+        let dst = asmt.dst.borrow();
+        match (&src.parent, &dst.parent) {
+            (
+                ir::PortParent::Cell(src_cell),
+                ir::PortParent::Cell(dst_cell),
+            ) => {
+                if src.attributes.has("done") && dst.attributes.has("go") {
+                    self.done_go = Some((
+                        src_cell.upgrade().borrow().clone_name(),
+                        dst_cell.upgrade().borrow().clone_name(),
+                    ));
+                }
+            }
+            // src_cell's done writes to group's done
+            (ir::PortParent::Cell(src_cell), ir::PortParent::Group(_)) => {
+                if dst.name == "done" && src.attributes.has("done") {
+                    self.last = Some(src_cell.upgrade().borrow().clone_name())
+                }
+            }
+            // If we encounter anything else, then not of interest to us
+            _ => (),
+        }
+    }
+
+    // Builds ordering for self. Returns true if this is a complete, valid,
+    // linear ordering in which all reads from the fist cell are from a
+    // stable port, false otherwise.
+    pub fn get_ordering(
+        &mut self,
+        asmts: &Vec<ir::Assignment>,
+    ) -> Option<(ir::Id, ir::Id)> {
+        //Update self.go_done_map and self.last for each asmt in the group.
+        for asmt in asmts {
+            self.update(asmt);
+        }
+        //Build ordering of cells, based on self.done_go and self.last.
+        if let (Some(last), Some((maybe_first, maybe_last))) =
+            (self.last.clone(), self.done_go.clone())
+        {
+            let all_stateful_writes =
+                ReadWriteSet::write_set(asmts.iter()).filter(comp_or_non_comb);
+            if maybe_last == last
+                // making sure maybe_fist and maybe_last are the only 2 cells written to 
+                && all_stateful_writes.count() == 2
+                // making sure that all reads of the first cell are from stable ports 
+                && asmts.iter().all(|assign| {
+                    if_name_stable_or_done(assign, &maybe_first)
+                })
+            {
+                Some((maybe_first, maybe_last))
+            } else {
```

Comment:
You can collapse some of these common `else { None }` branches by making the `if` branch `return Some(..);` and just have a default return `None` at the end of the function